### PR TITLE
[feature](Nereids)support join algorithm in Nereids that already implemented in original planner

### PR DESCRIPTION
**Pull request by morrySnow**
# Proposed changes

This PR
1. add support below join algorithm already supported by legacy to Nereids:
- colocate join
- bucket shuffle join
- shuffle join
- broadcast join

2. fix all cost enforce derive
- ChildOutputPropertyDeriver
- EnforceMissingPropertiesHelper
- RequestPropertyDeriver

3. add a local quick sort plan used in enforce
4. set PhysicalProperties to PhysicalPlan when choose best plan from memo
5. rename Job#pushTask to Job#pushJob

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [x] No
    - [ ] I don't know
2. Has unit tests been added:
    - [x] Yes
    - [ ] No
    - [ ] No Need
7. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
8. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
9. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java [hunk 1]
```diff
@@ -169,35 +175,39 @@ public static Pair<List<Expression>, List<Expression>> extractExpressionForHashT
...
-        Pair<List<SlotReference>, List<SlotReference>> childSlots =
+    public static Pair<List<ExprId>, List<ExprId>> getOnClauseUsedSlots(
+                AbstractPhysicalJoin<? extends Plan, ? extends Plan> join) {
+        Pair<List<ExprId>, List<ExprId>> childSlotsExprId =
                 Pair.of(Lists.newArrayList(), Lists.newArrayList());
 
         List<SlotReference> leftSlots = Utils.getOutputSlotReference(join.left());
         List<SlotReference> rightSlots = Utils.getOutputSlotReference(join.right());
-        List<EqualTo> equalToList = join.getHashJoinConjuncts().stream()
+        List<EqualTo> equalToList =  join.getHashJoinConjuncts().stream()
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:185]: strange things.
Why `checkstyle` doesn't check this space.

Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/util/JoinUtils.java:185]: add a rule to check it in another PR #12354

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalDistribute.java [hunk 2]
```diff
@@ -35,30 +36,40 @@
 /**
  * Enforcer plan.
  */
-public class PhysicalDistribution<CHILD_TYPE extends Plan> extends PhysicalUnary<CHILD_TYPE> {
+public class PhysicalDistribute<CHILD_TYPE extends Plan> extends PhysicalUnary<CHILD_TYPE> {
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/physical/PhysicalDistribute.java:39]: we need verb. all logic algebra operator except relation is verb. such as aggregate, project.

**Review by jackwener — DISMISSED** [review]
LGTM

**Review by jackwener — APPROVED** [review]

**Pull request merged**